src/coreclr/dlls/mscordbi/CMakeLists.txt

```diff
@@ -109,6 +109,7 @@ elseif(CLR_CMAKE_HOST_UNIX)
 
     if(CLR_CMAKE_HOST_LINUX)
         add_dependencies(mscordbi pal_redefines_file)
+        set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -Wl,--undefined-version")
```

Comment:
Clang 16 has flipped the switch on undefined symbol files to help users identify undefined symbols: https://github.com/llvm/llvm-project/commit/241dbd310599e3c1a0f1b0c9ced14c8b8760539e

It was throwing:

```
ld.lld: error: version script assignment of 'V1.0' to symbol 'PAL_RegisterModule' failed: symbol not defined
ld.lld: error: version script assignment of 'V1.0' to symbol 'PAL_UnregisterModule' failed: symbol not defined
```
when linking libmscordbi.so. With older `lld.ld` versions  `-Wl,--no-undefined-version` gives the same error. Which means `PAL_RegisterModule` and `PAL_UnregisterModule` are undefined in mscordbi since we have switched to `lld`.

cc @janvorli, should we remove them from the exports list for mscordbi?

Reply:
I believe so, @mikem8361, can you please confirm that?

Reply:
Thanks! I have pushed 6722dc0 for CI testing (which can be reverted). I will wait for @mikem8361's confirmation.

Reply:
What is suppose to happen is the libmscordbi needs to export these PAL functions from (and this is the broken part) the PAL in its libmscordacore link/reference. The dbi module doesn't have its own PAL linked in; it uses the dac module's PAL.  I don't know how to fix this.  The ld error makes it clear that the dbi export wasn't properly happening. 

These 2 PAL functions still need to be exported from dbi so the PAL load module function (via LOADRegisterLibraryDirect)  can create a module handle from the loaded module's PAL (i.e. libmscordbi) so the loading (i.e. libdbgshim) module can call the GetProcAddress on dbi.  libdbgshim has been moved out of band to the diagnostics repo so it isn't as clear anymore who is doing this PAL load library.

Reply:
The DBI exports might be broken because the DAC PAL exports are decorated with a DAC_ prefix. A "thunk" like this should fix the error:

```
PALIMPORT HINSTANCE PALAPI DAC_PAL_RegisterModule(IN LPCSTR lpLibFileName);
PALIMPORT VOID PALAPI DAC_PAL_UnregisterModule(IN HINSTANCE hInstance);

HINSTANCE
PALAPI
PAL_RegisterModule(
    IN LPCSTR lpLibFileName)
{
     return DAC_PAL_RegisterModule(lpLibFileName);
}

VOID
PALAPI
PAL_UnregisterModule(
    IN HINSTANCE hInstance)
{
    DAC_PAL_UnregisterModule(hInstance);
}
```

Reply:
@mikem8361, thank you for the explanation. I think the thunk in question should be generated by cmake: https://github.com/dotnet/runtime/blob/3ced9681d9eda35cee4c580066ecb3c10a569a1f/src/coreclr/dlls/mscordac/CMakeLists.txt#L58 which invokes a shell script that skips all the lines in exports list which do not start with `#`: https://github.com/dotnet/runtime/blob/3ced9681d9eda35cee4c580066ecb3c10a569a1f/src/coreclr/generateredefinesfile.sh#L29

I pushed a commit which fixes clang-16 build just by prepending `PAL_{Unr,R}egisterModule` lines with `#`: https://github.com/dotnet/runtime/pull/81573/commits/3542b1d3f01981c378ae5dbac7d3252dceea8037. Additionally, I have also removed this strange workaround https://github.com/dotnet/runtime/pull/81573/commits/9ada7b8b35177c43bf002785a015c586e5e3adf0 (which was supposedly fixing the hidden symbol issue by double lookup; doesn't seem to make any difference with lld as well as ld).